service/storage_service.cc

```diff
@@ -1536,6 +1536,39 @@ future<> storage_service::update_topology_with_local_metadata(raft::server& raft
     co_await _sys_ks.local().set_must_synchronize_topology(false);
 }
 
+// Wait until gossiper observes the given nodes or timeout is reached.
+// Called on initial startup (bootstrap/replace), not during restarts.
+static future<> wait_until_seen_during_boot(
+        const gms::gossiper& g, const std::unordered_set<gms::inet_address>& initial_contact_nodes,
+        db::timeout_clock::time_point timeout, abort_source& as) {
+    logger::rate_limit rate_limit{std::chrono::seconds{5}};
+    while (!as.abort_requested()) {
+        std::vector<gms::inet_address> observed = g.get_endpoints();
+        std::unordered_set<gms::inet_address> observed_set{observed.begin(), observed.end()};
```

Comment:
Nitpick: Instead of copying the list twice, I think you can use the zero-copy g.get_endpoints_states - it's a map, not a set, but I think its keys are what you need.
Of course the performance impact of this boot-time stuff is negligible, so I don't insist, but maybe it will even make the code a bit shorter.